### PR TITLE
extmod/re1.5: Fix escaping in RE classes.

### LINKED ISSUE #3178: "ure" incorrectly handles escaped '-' and trailing literal '-' in character classes
According to [the CPython docs](https://docs.python.org/3/library/re.html#regular-expression-syntax), in a character class,
>  If `-` is escaped (e.g. `[a\-z]`) or if it’s placed as the first or last character (e.g. `[a-]`), it will match a literal `'-'`.

---
The first assertion statement below succeeds, but the next four fail. As indicated, some of the failures are AssertionErrors (incorrect result was returned) and some are ValueErrors (input was rejected, no value returned).

```python
import ure

# Passing:
assert ure.compile(r'[-a]').split('foo-bar') == ['foo', 'b', 'r']

# AssertionError:
assert ure.compile(r'[a\-x]').split('foo-bar') == ['foo', 'b', 'r']
assert ure.compile(r'[\-ax]').split('foo-bar') == ['foo', 'b', 'r']

# ValueError: Error in regex
assert ure.compile(r'[ax\-]').split('foo-bar') == ['foo', 'b', 'r']
assert ure.compile(r'[a-]').split('foo-bar') == ['foo', 'b', 'r']
```

All 5 of these work fine in CPython with `re` and in MicroPython with re-pcre.

### PR DESCRIPTION
**Pull request by jimmo**
Fixes #3178 and #5220

Adds tests, including all the cases mentioned in both bugs.

### CODE REVIEW

File: extmod/re1.5/compilecode.c [hunk 1]
```diff
@@ -54,6 +54,10 @@ static const char *_compilecode(const char *re, ByteProg *prog, int sizecode)
             prog->len++;
             for (cnt = 0; *re != ']'; re++, cnt++) {
                 if (!*re) return NULL;
+                if (*re == '\\') {
+                    ++re;
+                    if (!*re) return NULL;
+                }
```
Comment (andrewleech) [inline extmod/re1.5/compilecode.c]: ok, that's much simpler to fix than I was expecting!

**Comment by andrewleech** [comment]
Looks good to me

File: extmod/re1.5/compilecode.c [hunk 2]
```diff
@@ -54,6 +54,10 @@ static const char *_compilecode(const char *re, ByteProg *prog, int sizecode)
             prog->len++;
             for (cnt = 0; *re != ']'; re++, cnt++) {
                 if (!*re) return NULL;
+                if (*re == '\\') {
+                    ++re;
+                    if (!*re) return NULL;
```
Comment (stinos) [inline extmod/re1.5/compilecode.c]: I know the rest of the file uses this style as well, but new code should probably adhere to the hous style i.e. use brackets?

**Comment by stinos** [comment]
Not sure what the status is of the code here vs where it originally came from (pfalcon's repository I guess?) but ideally this gets fixed there as well?

File: extmod/re1.5/compilecode.c (hunk 2 above)
Comment (andrewleech) [reply extmod/re1.5/compilecode.c]: Considering this file is from an external module, I would think it best to maintain the existing style.

Comment (stinos) [reply extmod/re1.5/compilecode.c]: Yeah, not sure, sort of depends on my other question about the relation to upstream, if any.

Anyway: the code can be simplified I think, move the increment before the existing NULL check and only one check is needed:

            for (cnt = 0; *re != ']'; re++, cnt++) {
                if (*re == '\\') {
                    ++re;
                }
                if (!*re) return NULL;

Comment (jimmo) [reply extmod/re1.5/compilecode.c]: Yes, good catch! Thanks

Regarding upstream... it already has diverged.

My intention was to match the existing code style.

**Comment by dpgeorge** [comment]
> Not sure what the status is of the code here vs where it originally came from (pfalcon's repository I guess?) but ideally this gets fixed there as well?

Support for escape within `[..]` was already implemented there, upstream.  But it was done in a different (and incompatible) way to what's done here, and didn't fix the cases in #3178, which is a nice benefit of the fix here.

Merged in ebf8332104c1c6fbd01b521a713e49d040c62920

**Pull request closed without merging**